### PR TITLE
Use Label-Looping algorithm for RNN-T decoding by default

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

Enable Label-Looping algorithm introduced in #8286 and #7926 (`loop_labels=True`) by default for RNN-T greedy decoding.

**Collection**: [ASR]

# Changelog 
- Enable Label-Looping algorithm by default (`loop_labels=true`)
- fix Label Looping algorithm for RNN-T/TDT + Stateless network
- fix tests with custom RNNT Decoder
- parametrize tests for batched greedy decoding to test both algorithms (Frame-/Label-Looping)

# Usage

Label-Looping algorithm is used by default now for batched greedy decoding.

For Frame-Looping algorithm one can use:
```shell
python examples/asr/speech_to_text_eval.py  <...> \
 rnnt_decoding.greedy.loop_labels=false
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
Not ready yet

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/modules/rnnt.py [hunk 1]
```diff
@@ -310,7 +310,9 @@ def score_hypothesis(
 
     def initialize_state(self, y: torch.Tensor) -> List[torch.Tensor]:
         batch = y.size(0)
-        state = [torch.ones([batch, self.context_size], dtype=torch.long, device=y.device) * self.blank_idx]
+        # state contains context_size - 1 elements for each utterance in batch,
+        # consistent with the state returned from StatelessNet.forward
+        state = [torch.ones([batch, self.context_size - 1], dtype=torch.long, device=y.device) * self.blank_idx]
```
Comment (artbataev) [inline nemo/collections/asr/modules/rnnt.py:315]: @hainan-xv, please, confirm that I broke nothing when fixing state for the Stateless decoder.
We need the state with the constant size (to allow replacements when we found the end of utterance), and `forward` returns the state of size `[batch_size, context_size - 1]`

File: tests/collections/asr/test_asr_rnnt_encdec_model.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def predict(
                 return (
                     output,
                     [
-                        torch.tensor([0] * self.vocab_size + [1], dtype=torch.float32)[None, None, :].exand(
+                        torch.tensor([0] * self.vocab_size + [1], dtype=torch.float32)[None, None, :].expand(
```
Comment (titu1994) [inline tests/collections/asr/test_asr_rnnt_encdec_model.py:76]: How how did this test pass with this error ?

Comment (artbataev) [reply tests/collections/asr/test_asr_rnnt_encdec_model.py:76]: I think this case is redundant and never executed in decoding, but we need to implement this to conform the interface where `y` is optional (see `AbstractRNNTDecoder.predict`)

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (galv) [reply nemo/collections/asr/modules/rnnt.py:315]: FYI, you can also use torch.full instead of torch.ones followed by multiplication. No need to change it though.

File: tests/collections/asr/test_asr_rnnt_encdec_model.py [hunk 3]
```diff
@@ -90,22 +90,25 @@ def predict(
...
-            return None
+        def initialize_state(self, y: torch.Tensor) -> List[torch.Tensor]:
+            batch_size = y.shape[0]
+            # NB: .clone is necessary after .expand, since the decoding algorithm manipulates the state
+            # (replacing elements), and this requires the state to be a real full tensor
+            # (not an expanded view, in which different elements can refer to the same memory location)
+            return [
+                torch.tensor([0] * self.vocab_size + [1], dtype=torch.float32)[None, None, :]
+                .expand([1, batch_size, -1])
+                .clone()
```
Comment (galv) [inline tests/collections/asr/test_asr_rnnt_encdec_model.py:101]: torch.repeat or torch.repeat_interleave is probably the better way to do it than expand followed by clone.

**Review by galv — COMMENTED** [review]
Late review with a few FYI comments. Do we test the cuda graphs implementation with stateless transducers yet?

**Comment by artbataev** [comment]
> Late review with a few FYI comments. Do we test the cuda graphs implementation with stateless transducers yet?

Thanks for the comments, @galv! I will make fixes in next PRs. 
Cuda graphs implementation with stateless is not tested on CI for now (no production model), but I tested loop labels + cuda graphs locally with a custom model, and it works.
